### PR TITLE
Feature/application

### PR DESCRIPTION
**Pull request by ccolombijn**
Give any comment or question that you might have ...

### CODE REVIEW

File: app/config.js [hunk 1]
```diff
@@ -3,16 +3,28 @@
...
+    //res.setHeader('Content-Type', 'application/json');
+   
+    ///const name = _config.name;
+    //const obj = _config.client;
+    _config.models ={}
+    for(let model of Object.getOwnPropertyNames(models.sequelize.models)){
+      _config.models[model] = models[model].rawAttributes
+    }
+    //_config.models = obj.models
+    //obj.version = _config.version;
```
Comment (rloman) [inline app/config.js:22]: Too much comments? Please remove uncommented code.

File: app/index.js [hunk 2]
```diff
@@ -1,16 +1,12 @@
...
-
-module.exports = (app)=>{
+module.exports = (app,config)=>{
     console.log('\x1b[32m[config.server.require]\x1b[0m \x1b[3mconfig/app.json\x1b[0m')
-    const requires = config.server.require;
-    for(const item of requires){
-        require(`./${item}`)(app);
+    for(const module of config.server.require){
+        require(`./${module}`)(app);
         //
```
Comment (rloman) [inline app/index.js:8]: Why this line?

File: assets/js/application-debug.js [hunk 3]
```diff
@@ -26,8 +26,16 @@ const debug = (function(){
...
             if(typeof configDebug === 'boolean'){
-                console.log(log)
-            
+                if(configDebug)console.log(log)
+            //}else if(configDebug.length){ // filter array
+            }else if( configDebug.constructor === Array){
+                //console.log('array')
+                for(let item in configDebug){
+                    
+                    if(log.indexOf(configDebug[item])>0){
```
Comment (rloman) [inline assets/js/application-debug.js:35]: indexOf returns -1 if not found
indexOf returns 0 or larger if found. If in this line the line starts with the term the result is false while it should be true???
"snake".indexOf("sna") will return in 0 (zero). Got it?

Comment (rloman) [reply assets/js/application-debug.js:35]: so might change to indexOf(...) >= 0 ?

File: assets/js/components.js [hunk 4]
```diff
@@ -0,0 +1,1027 @@
...
+    date : date,
+    time : time,
+    uid : uid,
+    auth : auth,
+    api : api
+  }
+  // .................................................
+  // component.date
+  function date(format){
+    const now = new Date();
```
Comment (rloman) [inline assets/js/components.js:53]: You might (later) using momentjs or even better use luxon.js (for now ok)

File: assets/js/components.js [hunk 5]
```diff
@@ -0,0 +1,1027 @@
...
+            .on('click',confirm); // confirm method of confirm object in arguments
+        $('.popover-body').append($confirmBtn)
+        const $cancelBtn = $('<button></button>')
+            .attr('class','btn btn-sm cancel')
+            .html('Cancel')
+            .on('click',cancel);
+        $('.popover-body').append($cancelBtn)
+      
+      }).on('hidden.bs.popover', () => { // popover is hidden
+          //button_.removeEventListener('click',confirm);
```
Comment (rloman) [inline assets/js/components.js:313]: What does this mean?

File: assets/js/components.js [hunk 6]
```diff
@@ -0,0 +1,1027 @@
...
+        form.appendChild(formRow(prop,args));
+      }else{
+        
+        if(prop.split('_')[0] === 'header'){
+          const header = document.createElement(prop.split('_')[1])
+          header.innerHTML = args.fields[prop]
+          const hr = document.createElement('hr')
+          form.appendChild(header);
+          form.appendChild(hr);
+        }else if(prop === 'field'){
```
Comment (rloman) [inline assets/js/components.js:691]: dubbele regels?????

File: assets/js/components.js [hunk 7]
```diff
@@ -0,0 +1,1027 @@
...
+    $documentFragment.append($calendarTable);
+    $(args.el).html($documentFragment);
+    return $documentFragment
+  }
+  // .................................................
+  // editor; all in one CRUD component
+  /*
+  component.editor({
+    module : 'modulename',
+    data : {
```
Comment (rloman) [inline assets/js/components.js:961]: too much uncommented code.

File: assets/js/components/type.js [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+            case Array:
+                callback(args);
+                break;
+            case Object:
+                callback(args);
+                break;
+            case Function:
+                callback(args);
+                break;
+
```
Comment (rloman) [inline assets/js/components/type.js:30]: Might add a default case here which might thrown an Error. Too prevent illegal prototypes to be forwarded here.

File: assets/js/modules/blockdates.js [hunk 9]
```diff
@@ -1,15 +1,77 @@
...
+    component.table({
+      model : 'Blockdate',
+      el: '#overviewTable',
+      data : blockdatesData,
+      class : 'table-striped table-hover',
+      cols : {
+        name : { label : 'Name' },
+        date : { label : 'Date' },
+        participants : { label : 'Participants' }
+      },
```
Comment (rloman) [inline assets/js/modules/blockdates.js:30]: too much commented code.

File: assets/js/modules/options.js [hunk 10]
```diff
@@ -3,20 +3,249 @@
...
+  optionsOverview = () => {
+   /* component.table({
+      model : 'Option',
+      el: '#optionsOverview',
+      data : optionsData,
+      class : 'table-striped table-hover',
+      cols : {
+        name : { label : 'Name' },
+        date : { label : 'Date' },
+        participants : { label : 'Participants' }
```
Comment (rloman) [inline assets/js/modules/options.js:169]: too much commented code

File: assets/js/utils.js [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+
+    return `${hh}:${mm}`;
+  }
+  // utils.datetime
+  function datetime(){
+      return `${date()} ${time()}`
+  }
+  // utils.uid
+  function uid(){  // generate unique id
+      const s4 = () => Math.floor((1 + Math.random()) * 0x10000).toString(16).substring(1);
```
Comment (rloman) [inline assets/js/utils.js:34]: are you sure this is ALWAYS unique???

File: controllers/accounts.js [hunk 12]
```diff
@@ -2,42 +2,170 @@
...
     Account.create(account).then((account)=>{
+        
         sgMail.setApiKey(process.env.SENDGRID_API_KEY);
         const msg = {
           to: `${account.email}`,
           from: `noreply@agendamanager.nl`,
           subject: `Registration at agendamanager.nl`,
           text: `Someone has invited you to join Agenda Manager. Visit agendamanager.nl/verify and paste the following code: ${account.id}`,
-          html: `${req.session.user.firstName} ${req.session.user.lastName} has invited you to join Agenda Manager. Visit <a href="${process.env.REF_URL}verify?uuid=${account.id}">agendamanager.nl/verify</a> and paste the following code: <br><strong>${account.id}</strong>`,
+          html: `<div><img src="https://dev.agendamanager.nl/img/logo_sm_300.png" style="margin:auto;" /></div>${req.session.user.firstName} ${req.session.user.lastName} has invited you to join Agenda Manager. Visit <a href="${process.env.REF_HTTP_PROTOCOL}://${process.env.REF_URL}verify?uuid=${account.id}">agendamanager.nl/verify</a> and paste the following code: <br><strong>${account.id}</strong>`,
```
Comment (rloman) [inline controllers/accounts.js:38]: Is dit plaatje altijd aanwezig??? (als de server niet draait dan werkt het wel???

**Review by rloman — APPROVED** [review]
Prima code, vooral veel en mooi geschreven. Maar let dus even op het belangrijkste: indexOf wordt soms met > 0 gebruikt en wellicht moet dit >= 0 zijn. En wat veel commented code ... verder prima werk, kerel!

File: assets/js/utils.js (hunk 11 above)
Comment (ccolombijn) [reply assets/js/utils.js:34]: Not really sure. See https://stackoverflow.com/questions/38228180/is-generating-and-concatenating-3-math-random-values-more-random-than-1-math-r

File: controllers/accounts.js (hunk 12 above)
Comment (ccolombijn) [reply controllers/accounts.js:38]: Is zeker geen definitieve oplossing, al zou de server we moeten draaien als de e-mail gelezen wordt (en het plaatje wordt getoond) ... anders werkt de bevestigingslink ook niet

File: assets/js/application-debug.js (hunk 3 above)
Comment (ccolombijn) [reply assets/js/application-debug.js:35]: Absolutely get your point, but in this particular case the string which is searched for is never at position 0 (function:method is logged; method needs to be filtered, not function), changed it tough.

File: assets/js/components.js (hunk 5 above)
Comment (ccolombijn) [reply assets/js/components.js:313]: not really anything ;) removed.

File: assets/js/utils.js (hunk 11 above)
Comment (ccolombijn) [reply assets/js/utils.js:34]: Added uuidv4 for as better solution

**Comment by ccolombijn** [comment]
Will add a new pull request next week as requested

**Pull request merged**